File: Knapsack.py
"""
Verdict: TLE
"""
from sys import stdin, stdout
def knapsack(c, n, values, weights):
    dp = [0] * (c+1)
    chosen = [[] for _ in range(c+1)]

    for i in range(n):
        for w in range(c, weights[i]-1, -1):
            if dp[w] < dp[w-weights[i]] + values[i]:
                dp[w] = dp[w-weights[i]] + values[i]
                chosen[w] = chosen[w-weights[i]] + [i]

    maxvalue = max(dp)
    maxidx = dp.index(maxvalue)

    return len(chosen[maxidx]), chosen[maxidx]

input = stdin.readline
print = stdout.write
while True:
    try:
        c, n = list(map(int, input().split()))
        values = []
        weights = []
        for _ in range(n):
            v, w = list(map(int, input().split()))
            values.append(v)
            weights.append(w)

        numChosen, chosen = knapsack(c, n, values, weights)
        print(f"{numChosen}\n")
        print(' '.join(map(str, chosen)) + '\n')
    
    except EOFError:
        break

File: Knapsack.cpp
#include <bits/stdc++.h>
#include <vector>
#include <algorithm>

using namespace std;

pair<int, vector<int>> knapsack(int capacity, int n, const vector<int>& values, const vector<int>& weights) {
    vector<vector<int>> dp(n+1, vector<int>(capacity + 1, 0));

    for(int i = 1; i <= n; i++) {
        for(int w = 1; w <= capacity; w++) {
            if (weights[i-1] <= w) dp[i][w] = max(values[i-1] + dp[i-1][w-weights[i-1]], dp[i-1][w]);
            else dp[i][w] = dp[i-1][w];
        }
    }

    vector<int> chosen;

    int w = capacity;
    for(int i = n; i > 0; i--) {
        if(dp[i][w] != dp[i-1][w]) {
            chosen.push_back(i-1);
            w -= weights[i-1];
        }
    }
    reverse(chosen.begin(), chosen.end());
    return {chosen.size(), chosen};
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int c, n;
    while (cin >> c >> n) {
        vector<int> values(n), weights(n);

        for(int i = 0; i < n; i++){
            cin >> values[i] >> weights[i];
        }

        auto [numchosen, chosen] = knapsack(c, n, values, weights);

        cout << numchosen << "\n";

        for(int i = 0; i < numchosen; i++) {
            cout << chosen[i] << (i < numchosen - 1 ? " " : "\n");
        }
    }

    return 0;
};